File: todo.js
// todo.js
#!/usr/bin/env node
'use strict';

const fs = require('fs');
const path = require('path');
const os = require('os');

const COLORS = {
    reset: '\x1b[0m',
    red: '\x1b[91m',
    green: '\x1b[92m',
    yellow: '\x1b[93m',
    blue: '\x1b[94m',
    magenta: '\x1b[95m',
    cyan: '\x1b[96m',
    bold: '\x1b[1m'
};

function colorize(text, color) {
    return COLORS[color] + text + COLORS.reset;
}

const TASKS_FILE = path.join(os.homedir(), '.todo_tasks.json');

function loadTasks() {
    try {
        return JSON.parse(fs.readFileSync(TASKS_FILE, 'utf8'));
    } catch {
        return [];
    }
}

function saveTasks(tasks) {
    fs.writeFileSync(TASKS_FILE, JSON.stringify(tasks, null, 2));
}

function getNextID(tasks) {
    let max = 0;
    for (const t of tasks) {
        if (t.id > max) max = t.id;
    }
    return max + 1;
}

function taskColor(task) {
    if (task.status === 'done') return 'green';
    switch (task.priority) {
        case 'High': return 'red';
        case 'Medium': return 'yellow';
        case 'Low': return 'blue';
        default: return 'reset';
    }
}

function prioritySymbol(task) {
    switch (task.priority) {
        case 'High': return '🔴';
        case 'Medium': return '🟡';
        case 'Low': return '🔵';
        default: return '⚪';
    }
}

function addTask(title, description, priority, deadline) {
    const tasks = loadTasks();
    const task = {
        id: getNextID(tasks),
        title,
        description: description || '',
        priority: priority || 'Medium',
        deadline: deadline || '',
        status: 'pending',
        created_at: new Date().toISOString()
    };
    tasks.push(task);
    saveTasks(tasks);
    console.log(colorize(`✅ Задача #${task.id} добавлена.`, 'green'));
}

function listTasks(filter, sortBy) {
    let tasks = loadTasks();
    if (tasks.length === 0) {
        console.log(colorize('Нет задач.', 'yellow'));
        return;
    }
    if (filter) {
        const [key, val] = filter.split('=');
        tasks = tasks.filter(t => t[key] === val);
    }
    if (sortBy === 'priority') {
        const order = { High: 0, Medium: 1, Low: 2 };
        tasks.sort((a, b) => order[a.priority] - order[b.priority]);
    } else if (sortBy === 'deadline') {
        tasks.sort((a, b) => (a.deadline || '9999-12-31').localeCompare(b.deadline || '9999-12-31'));
    } else if (sortBy === 'created') {
        tasks.sort((a, b) => a.created_at.localeCompare(b.created_at));
    }
    for (const t of tasks) {
        const status = t.status === 'done' ? '✅' : '⬜';
        const deadline = t.deadline ? ` до ${t.deadline}` : '';
        const priorityColored = colorize(t.priority, taskColor(t));
        console.log(`[${t.id}] ${status} ${t.title} ${priorityColored} ${deadline}`);
    }
}

function doneTask(id) {
    const tasks = loadTasks();
    const task = tasks.find(t => t.id === id);
    if (!task) {
        console.log(colorize(`Задача #${id} не найдена.`, 'red'));
        return;
    }
    if (task.status === 'done') {
        console.log(colorize(`Задача #${id} уже выполнена.`, 'yellow'));
        return;
    }
    task.status = 'done';
    saveTasks(tasks);
    console.log(colorize(`✅ Задача #${id} отмечена как выполненная.`, 'green'));
}

function deleteTask(id) {
    let tasks = loadTasks();
    const idx = tasks.findIndex(t => t.id === id);
    if (idx === -1) {
        console.log(colorize(`Задача #${id} не найдена.`, 'red'));
        return;
    }
    tasks.splice(idx, 1);
    saveTasks(tasks);
    console.log(colorize(`🗑️ Задача #${id} удалена.`, 'yellow'));
}

function editTask(id, title, description, priority, deadline) {
    const tasks = loadTasks();
    const task = tasks.find(t => t.id === id);
    if (!task) {
        console.log(colorize(`Задача #${id} не найдена.`, 'red'));
        return;
    }
    if (title) task.title = title;
    if (description) task.description = description;
    if (priority) task.priority = priority;
    if (deadline) task.deadline = deadline;
    saveTasks(tasks);
    console.log(colorize(`✅ Задача #${id} обновлена.`, 'green'));
}

function searchTasks(query) {
    const tasks = loadTasks();
    const results = tasks.filter(t =>
        t.title.toLowerCase().includes(query.toLowerCase()) ||
        t.description.toLowerCase().includes(query.toLowerCase())
    );
    if (results.length === 0) {
        console.log(colorize(`Ничего не найдено по запросу '${query}'.`, 'yellow'));
        return;
    }
    console.log(colorize(`🔍 Найдено ${results.length} задач:`, 'bold'));
    for (const t of results) {
        console.log(`  [${t.id}] ${t.title}`);
    }
}

function stats() {
    const tasks = loadTasks();
    const total = tasks.length;
    const done = tasks.filter(t => t.status === 'done').length;
    const pending = total - done;
    console.log(colorize('📊 Статистика:', 'bold'));
    console.log(`  Всего задач: ${total}`);
    if (total > 0) {
        console.log(`  Выполнено: ${done} (${(done/total*100).toFixed(1)}%)`);
    } else {
        console.log('  Выполнено: 0');
    }
    console.log(`  Ожидают: ${pending}`);
}

function exportTasks(filename) {
    const tasks = loadTasks();
    fs.writeFileSync(filename, JSON.stringify(tasks, null, 2));
    console.log(colorize(`✅ Задачи экспортированы в ${filename}`, 'green'));
}

function importTasks(filename) {
    const data = JSON.parse(fs.readFileSync(filename, 'utf8'));
    const tasks = loadTasks();
    for (const t of data) {
        t.id = getNextID(tasks);
        tasks.push(t);
    }
    saveTasks(tasks);
    console.log(colorize(`✅ Импортировано ${data.length} задач из ${filename}`, 'green'));
}

function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.log(`Usage: node todo.js <add|list|done|delete|edit|search|stats|export|import> [args...]`);
        console.log(`  add <title> [description] -p <priority> -d <deadline>`);
        console.log(`  list [--filter status=pending|done] [--sort priority|deadline|created]`);
        console.log(`  done <id>`);
        console.log(`  delete <id>`);
        console.log(`  edit <id> --title <title> --description <desc> --priority <p> --deadline <d>`);
        console.log(`  search <query>`);
        console.log(`  stats`);
        console.log(`  export <file>`);
        console.log(`  import <file>`);
        process.exit(1);
    }
    const cmd = args[0];

    switch (cmd) {
        case 'add': {
            if (args.length < 2) {
                console.log('Использование: add <title> [description] -p <priority> -d <deadline>');
                return;
            }
            let title = args[1];
            let description = '';
            let priority = 'Medium';
            let deadline = '';
            for (let i = 2; i < args.length; i++) {
                if (args[i] === '-p' && i+1 < args.length) {
                    priority = args[++i];
                } else if (args[i] === '-d' && i+1 < args.length) {
                    deadline = args[++i];
                } else if (!description) {
                    description = args[i];
                }
            }
            addTask(title, description, priority, deadline);
            break;
        }
        case 'list': {
            let filter = '';
            let sortBy = '';
            for (let i = 1; i < args.length; i++) {
                if (args[i] === '--filter' && i+1 < args.length) {
                    filter = args[++i];
                } else if (args[i] === '--sort' && i+1 < args.length) {
                    sortBy = args[++i];
                }
            }
            listTasks(filter, sortBy);
            break;
        }
        case 'done': {
            if (args.length < 2) {
                console.log('Использование: done <id>');
                return;
            }
            doneTask(parseInt(args[1]));
            break;
        }
        case 'delete': {
            if (args.length < 2) {
                console.log('Использование: delete <id>');
                return;
            }
            deleteTask(parseInt(args[1]));
            break;
        }
        case 'edit': {
            if (args.length < 3) {
                console.log('Использование: edit <id> [--title <title>] [--description <desc>] [--priority <p>] [--deadline <d>]');
                return;
            }
            const id = parseInt(args[1]);
            let title = '', description = '', priority = '', deadline = '';
            for (let i = 2; i < args.length; i++) {
                if (args[i] === '--title' && i+1 < args.length) {
                    title = args[++i];
                } else if (args[i] === '--description' && i+1 < args.length) {
                    description = args[++i];
                } else if (args[i] === '--priority' && i+1 < args.length) {
                    priority = args[++i];
                } else if (args[i] === '--deadline' && i+1 < args.length) {
                    deadline = args[++i];
                }
            }
            editTask(id, title, description, priority, deadline);
            break;
        }
        case 'search': {
            if (args.length < 2) {
                console.log('Использование: search <query>');
                return;
            }
            searchTasks(args[1]);
            break;
        }
        case 'stats':
            stats();
            break;
        case 'export': {
            if (args.length < 2) {
                console.log('Использование: export <file>');
                return;
            }
            exportTasks(args[1]);
            break;
        }
        case 'import': {
            if (args.length < 2) {
                console.log('Использование: import <file>');
                return;
            }
            importTasks(args[1]);
            break;
        }
        default:
            console.log('Неизвестная команда:', cmd);
    }
}

main();

File: todo.rb
#!/usr/bin/env ruby
# todo.rb
# encoding: UTF-8

require 'json'
require 'time'
require 'optparse'
require 'fileutils'

COLORS = {
  reset: "\e[0m",
  red: "\e[91m",
  green: "\e[92m",
  yellow: "\e[93m",
  blue: "\e[94m",
  bold: "\e[1m"
}

def colorize(text, color)
  "#{COLORS[color]}#{text}#{COLORS[:reset]}"
end

TASKS_FILE = File.join(Dir.home, '.todo_tasks.json')

def load_tasks
  return [] unless File.exist?(TASKS_FILE)
  JSON.parse(File.read(TASKS_FILE))
end

def save_tasks(tasks)
  File.write(TASKS_FILE, JSON.pretty_generate(tasks))
end

def next_id(tasks)
  tasks.empty? ? 1 : tasks.map { |t| t['id'] }.max + 1
end

def task_color(task)
  return 'green' if task['status'] == 'done'
  case task['priority']
  when 'High' then 'red'
  when 'Medium' then 'yellow'
  when 'Low' then 'blue'
  else 'reset'
  end
end

def priority_symbol(task)
  case task['priority']
  when 'High' then '🔴'
  when 'Medium' then '🟡'
  when 'Low' then '🔵'
  else '⚪'
  end
end

def add_task(title, description, priority, deadline)
  tasks = load_tasks
  task = {
    'id' => next_id(tasks),
    'title' => title,
    'description' => description || '',
    'priority' => priority || 'Medium',
    'deadline' => deadline || '',
    'status' => 'pending',
    'created_at' => Time.now.iso8601
  }
  tasks << task
  save_tasks(tasks)
  puts colorize("✅ Задача ##{task['id']} добавлена.", :green)
end

def list_tasks(filter, sort_by)
  tasks = load_tasks
  if tasks.empty?
    puts colorize('Нет задач.', :yellow)
    return
  end
  unless filter.nil?
    key, val = filter.split('=')
    tasks = tasks.select { |t| t[key] == val }
  end
  case sort_by
  when 'priority'
    order = { 'High' => 0, 'Medium' => 1, 'Low' => 2 }
    tasks.sort_by! { |t| order[t['priority']] }
  when 'deadline'
    tasks.sort_by! { |t| t['deadline'] || '9999-12-31' }
  when 'created'
    tasks.sort_by! { |t| t['created_at'] }
  end
  tasks.each do |t|
    status = t['status'] == 'done' ? '✅' : '⬜'
    deadline = t['deadline'].empty? ? '' : " до #{t['deadline']}"
    puts "[#{t['id']}] #{status} #{t['title']} #{colorize(t['priority'], task_color(t))} #{deadline}"
  end
end

def done_task(id)
  tasks = load_tasks
  task = tasks.find { |t| t['id'] == id }
  unless task
    puts colorize("Задача ##{id} не найдена.", :red)
    return
  end
  if task['status'] == 'done'
    puts colorize("Задача ##{id} уже выполнена.", :yellow)
    return
  end
  task['status'] = 'done'
  save_tasks(tasks)
  puts colorize("✅ Задача ##{id} отмечена как выполненная.", :green)
end

def delete_task(id)
  tasks = load_tasks
  task = tasks.find { |t| t['id'] == id }
  unless task
    puts colorize("Задача ##{id} не найдена.", :red)
    return
  end
  tasks.delete(task)
  save_tasks(tasks)
  puts colorize("🗑️ Задача ##{id} удалена.", :yellow)
end

def edit_task(id, title, description, priority, deadline)
  tasks = load_tasks
  task = tasks.find { |t| t['id'] == id }
  unless task
    puts colorize("Задача ##{id} не найдена.", :red)
    return
  end
  task['title'] = title unless title.nil?
  task['description'] = description unless description.nil?
  task['priority'] = priority unless priority.nil?
  task['deadline'] = deadline unless deadline.nil?
  save_tasks(tasks)
  puts colorize("✅ Задача ##{id} обновлена.", :green)
end

def search_tasks(query)
  tasks = load_tasks
  results = tasks.select do |t|
    t['title'].downcase.include?(query.downcase) ||
    t['description'].downcase.include?(query.downcase)
  end
  if results.empty?
    puts colorize("Ничего не найдено по запросу '#{query}'.", :yellow)
    return
  end
  puts colorize("🔍 Найдено #{results.size} задач:", :bold)
  results.each { |t| puts "  [#{t['id']}] #{t['title']}" }
end

def stats
  tasks = load_tasks
  total = tasks.size
  done = tasks.count { |t| t['status'] == 'done' }
  pending = total - done
  puts colorize("📊 Статистика:", :bold)
  puts "  Всего задач: #{total}"
  if total > 0
    puts "  Выполнено: #{done} (#{(done.to_f/total*100).round(1)}%)"
  else
    puts "  Выполнено: 0"
  end
  puts "  Ожидают: #{pending}"
end

def export_tasks(filename)
  tasks = load_tasks
  File.write(filename, JSON.pretty_generate(tasks))
  puts colorize("✅ Задачи экспортированы в #{filename}", :green)
end

def import_tasks(filename)
  unless File.exist?(filename)
    puts colorize("Файл не найден.", :red)
    return
  end
  imported = JSON.parse(File.read(filename))
  tasks = load_tasks
  imported.each do |t|
    t['id'] = next_id(tasks)
    tasks << t
  end
  save_tasks(tasks)
  puts colorize("✅ Импортировано #{imported.size} задач из #{filename}", :green)
end

def main
  if ARGV.empty?
    puts "Usage: todo.rb <add|list|done|delete|edit|search|stats|export|import> [args...]"
    puts "  add <title> [description] -p <priority> -d <deadline>"
    puts "  list [--filter status=pending|done] [--sort priority|deadline|created]"
    puts "  done <id>"
    puts "  delete <id>"
    puts "  edit <id> --title <title> --description <desc> --priority <p> --deadline <d>"
    puts "  search <query>"
    puts "  stats"
    puts "  export <file>"
    puts "  import <file>"
    exit 1
  end

  cmd = ARGV[0]
  case cmd
  when 'add'
    if ARGV.size < 2
      puts "Использование: add <title> [description] -p <priority> -d <deadline>"
      return
    end
    title = ARGV[1]
    description = ''
    priority = 'Medium'
    deadline = ''
    i = 2
    while i < ARGV.size
      case ARGV[i]
      when '-p' then priority = ARGV[i+1]; i += 1
      when '-d' then deadline = ARGV[i+1]; i += 1
      else description = ARGV[i] if description.empty?
      end
      i += 1
    end
    add_task(title, description, priority, deadline)

  when 'list'
    filter = nil
    sort_by = nil
    i = 1
    while i < ARGV.size
      case ARGV[i]
      when '--filter' then filter = ARGV[i+1]; i += 1
      when '--sort' then sort_by = ARGV[i+1]; i += 1
      end
      i += 1
    end
    list_tasks(filter, sort_by)

  when 'done'
    if ARGV.size < 2
      puts "Использование: done <id>"
      return
    end
    done_task(ARGV[1].to_i)

  when 'delete'
    if ARGV.size < 2
      puts "Использование: delete <id>"
      return
    end
    delete_task(ARGV[1].to_i)

  when 'edit'
    if ARGV.size < 3
      puts "Использование: edit <id> --title <title> --description <desc> --priority <p> --deadline <d>"
      return
    end
    id = ARGV[1].to_i
    title = description = priority = deadline = nil
    i = 2
    while i < ARGV.size
      case ARGV[i]
      when '--title' then title = ARGV[i+1]; i += 1
      when '--description' then description = ARGV[i+1]; i += 1
      when '--priority' then priority = ARGV[i+1]; i += 1
      when '--deadline' then deadline = ARGV[i+1]; i += 1
      end
      i += 1
    end
    edit_task(id, title, description, priority, deadline)

  when 'search'
    if ARGV.size < 2
      puts "Использование: search <query>"
      return
    end
    search_tasks(ARGV[1])

  when 'stats'
    stats

  when 'export'
    if ARGV.size < 2
      puts "Использование: export <file>"
      return
    end
    export_tasks(ARGV[1])

  when 'import'
    if ARGV.size < 2
      puts "Использование: import <file>"
      return
    end
    import_tasks(ARGV[1])

  else
    puts "Неизвестная команда: #{cmd}"
  end
end

main if __FILE__ == $0
